File: lib/java/src/main/java/uapi/Serializer.java
package uapi;

import static uapi.internal.DeserializeInternal.deserializeInternal;
import static uapi.internal.SerializeInternal.serializeInternal;

import uapi.internal.binary.BinaryEncoder;

/**
 * A serializer that converts a Message to and from a serialized form.
 */
public class Serializer {

    private Serialization serializationImpl;
    private BinaryEncoder binaryEncoder;

    Serializer(Serialization serializationImpl, BinaryEncoder binaryEncoder) {
        this.serializationImpl = serializationImpl;
        this.binaryEncoder = binaryEncoder;
    }

    /**
     * Serialize a Message into a byte array.
     * 
     * @param message
     * @return
     */
    public byte[] serialize(Message message) {
        return serializeInternal(message, this.binaryEncoder, this.serializationImpl);
    }

    /**
     * Deserialize a Message from a byte array.
     * 
     * @param messageBytes
     * @return
     */
    public Message deserialize(byte[] messageBytes) {
        return deserializeInternal(messageBytes, this.serializationImpl, this.binaryEncoder);
    }
}

File: lib/py/uapi/Serializer.py
from typing import Any

from uapi.Message import Message
from uapi.Serialization import Serialization
from uapi.internal.DeserializeInternal import deserialize_internal
from uapi.internal.SerializeInternal import serialize_internal
from uapi.internal.binary.BinaryEncoder import BinaryEncoder


class Serializer:
    """
    A serializer that converts a Message to and from a serialized form.
    """

    def __init__(self, serialization_impl: Serialization, binary_encoder: BinaryEncoder):
        self.serialization_impl = serialization_impl
        self.binary_encoder = binary_encoder

    def serialize(self, message: Message) -> bytes:
        """
        Serialize a Message into a byte array.
        """
        return serialize_internal(message, self.binary_encoder, self.serialization_impl)

    def deserialize(self, message_bytes: bytes) -> Message:
        """
        Deserialize a Message from a byte array.
        """
        return deserialize_internal(message_bytes, self.serialization_impl, self.binary_encoder)
